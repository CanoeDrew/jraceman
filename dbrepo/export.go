package dbrepo

import (
  "fmt"
  "io"
  "strings"

  "github.com/jimmc/jracemango/dbrepo/structsql"
)

func (r *Repos) Export(w io.Writer) error {
  if err := r.exportHeader(w); err != nil {
    return err
  }

  // The order of output of the tables is important: tables with
  // foreign keys should be after the tables the point to.

  if err := r.dbSite.Export(r, w); err != nil {
    return err
  }

  if err := r.dbArea.Export(r, w); err != nil {
    return err
  }

  return nil
}

func (r *Repos) exportHeader(w io.Writer) error {
  io.WriteString(w, "#!jraceman -import\n")
  io.WriteString(w, "!exportVersion 2\n")
  io.WriteString(w, "!appInfo JRaceman v2.0.0\n")       // TODO - get real version
  io.WriteString(w, "!type database\n")
  return nil
}

func (r *Repos) exportTableFromStruct(w io.Writer, tableName string, element interface{}) error {
  if err := r.exportTableHeaderFromStruct(w, tableName, element); err != nil {
    return err
  }
  return r.exportTableDataFromStruct(w, tableName, element)
}

func (r *Repos) exportTableHeaderFromStruct(w io.Writer, tableName string, element interface{}) error {
  io.WriteString(w, "\n!table " + tableName + "\n")
  colnames := `"` + strings.Join(structsql.ColumnNames(element), `","`) + `"`
  io.WriteString(w, "!columns " + colnames + "\n")
  return nil
}

func (r *Repos) exportTableDataFromStruct(w io.Writer, tableName string, element interface{}) error {
  sql, targets := structsql.SelectSql(tableName, element)
  sql = sql + ";"
  rows, err := r.db.Query(sql)
  if err != nil {
    return err
  }
  defer rows.Close()
  rowCount := 0
  for rows.Next() {
    err := rows.Scan(targets...)
    if err != nil {
      return err
    }
    // All of the targets are pointers to fields of the struct,
    // so if the struct field is a pointer (to allow null values),
    // then the type of the target is a double-pointer.
    for i, target := range targets {
      if i > 0 {
        io.WriteString(w, ",")
      }
      switch t := target.(type) {
      case *int:
        fmt.Fprintf(w, "%d", *t)
      case *string:
        fmt.Fprintf(w, "%q", *t)
      case **string:
        if *t == nil {
          fmt.Fprintf(w, "null")
        } else {
          fmt.Fprintf(w, "%q", **t)
        }
      default:
        // If we don't understand the type, print it out so that we know what
        // we need to add to the above switch statement.
        fmt.Fprintf(w, "(Type %T)", target)
      }
    }
    io.WriteString(w, "\n")
    rowCount++
  }
  if rowCount == 0 {
    io.WriteString(w, "#no rows\n")
  }
  return rows.Err()
}
