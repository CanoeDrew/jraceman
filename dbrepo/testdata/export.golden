#!jraceman -import
!exportVersion 2
!appInfo JRaceman v2.0.0
!type database

!table competition
!columns "id","name","groupsize","maxalternates","duration"
#no rows

!table complan
!columns "id","system","plan","minentries","maxentries","planorder"
#no rows

!table complanstage
!columns "id","complanid","stageid","round","sectioncount","fillorder"
#no rows

!table complanrule
!columns "id","complanid","fromround","fromsection","fromplace","toround","tosection","tolane"
#no rows

!table site
!columns "id","name","street","street2","city","state","zip","country","phone","fax"
"S1","Site One",null,null,"Anytown",null,null,null,null,null
"S2","Site Two",null,null,"Centerville",null,null,null,null,null

!table area
!columns "id","name","siteid","lanes","extralanes"
"A1","Main Course","S1",9,2
"A2","Distance Course","S1",150,0
"A3","Center Course","S2",9,0

!table exception
!columns "id","name","shortname","resultallowedrequired"
#no rows

!table level
!columns "id","name","minentryage","minage","maxage","maxentryage","usegroupaverage"
#no rows

!table stage
!columns "id","name","number","isfinal"
#no rows

!table gender
!columns "id","name"
#no rows

!table simplan
!columns "id","system","plan","minentries","maxentries"
#no rows

!table simplanstage
!columns "id","simplanid","stageid","sectioncount","fillorder"
#no rows

!table simplanrule
!columns "id","simplanid","fromstageid","tostageid","thruplace","nextbesttimes"
#no rows

!table progression
!columns "id","name","class","parameters"
#no rows

!table scoringsystem
!columns "id","name"
#no rows

!table scoringrule
!columns "id","scoringsystemid","rule","value","points"
#no rows
